File: android/src/main/kotlin/com/example/isometrik_chat_flutter/DeliveryReceiptApiClient.kt
package com.example.isometrik_chat_flutter

import android.content.Context
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/**
 * Native HTTP client for the existing SDK delivered-indicator API:
 * `PUT {baseUrl}/chat/indicator/delivered`
 */
object DeliveryReceiptApiClient {
    private const val DELIVERED_PATH = "/chat/indicator/delivered"
    private const val TIMEOUT_MS = 20_000

    /**
     * @return true when the server accepted the delivery ack (2xx).
     */
    fun ackDelivered(
        context: Context,
        messageId: String,
        conversationId: String,
    ): Boolean {
        if (!DeliveryReceiptConfigStore.isConfigured(context)) return false
        if (messageId.isBlank() || conversationId.isBlank()) return false

        val baseUrl = DeliveryReceiptConfigStore.baseUrl(context) ?: return false
        val url = URL("$baseUrl$DELIVERED_PATH")
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "PUT"
            connectTimeout = TIMEOUT_MS
            readTimeout = TIMEOUT_MS
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty(
                "licenseKey",
                DeliveryReceiptConfigStore.licenseKey(context) ?: "",
            )
            setRequestProperty(
                "appSecret",
                DeliveryReceiptConfigStore.appSecret(context) ?: "",
            )
            setRequestProperty(
                "userToken",
                DeliveryReceiptConfigStore.userToken(context) ?: "",
            )
        }

        return try {
            val body = JSONObject()
                .put("messageId", messageId)
                .put("conversationId", conversationId)
                .toString()

            OutputStreamWriter(connection.outputStream, Charsets.UTF_8).use { writer ->
                writer.write(body)
            }

            val code = connection.responseCode
            code in 200..299
        } catch (_: Exception) {
            false
        } finally {
            connection.disconnect()
        }
    }
}

File: ios/Classes/DeliveryReceiptApiClient.swift
import Foundation

/// Native client for the existing SDK delivered-indicator API:
/// `PUT {baseUrl}/chat/indicator/delivered`
enum DeliveryReceiptApiClient {
    private static let deliveredPath = "/chat/indicator/delivered"
    private static let timeout: TimeInterval = 20

    @discardableResult
    static func ackDelivered(
        messageId: String,
        conversationId: String,
        appGroupId: String?
    ) -> Bool {
        guard
            let config = DeliveryReceiptConfigStore.config(appGroupId: appGroupId),
            let baseUrl = config["baseUrl"],
            let userToken = config["userToken"],
            !messageId.isEmpty,
            !conversationId.isEmpty,
            let url = URL(string: baseUrl + deliveredPath)
        else { return false }

        var request = URLRequest(url: url, timeoutInterval: timeout)
        request.httpMethod = "PUT"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(config["licenseKey"] ?? "", forHTTPHeaderField: "licenseKey")
        request.setValue(config["appSecret"] ?? "", forHTTPHeaderField: "appSecret")
        request.setValue(userToken, forHTTPHeaderField: "userToken")

        let body: [String: String] = [
            "messageId": messageId,
            "conversationId": conversationId,
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        let semaphore = DispatchSemaphore(value: 0)
        var success = false

        URLSession.shared.dataTask(with: request) { _, response, _ in
            if let http = response as? HTTPURLResponse {
                success = (200...299).contains(http.statusCode)
            }
            semaphore.signal()
        }.resume()

        _ = semaphore.wait(timeout: .now() + timeout + 1)
        return success
    }
}
